src/dtx/dtx_common.c

```diff
@@ -1697,50 +1700,53 @@ struct dtx_ult_arg {
 static void
 dtx_leader_exec_ops_ult(void *arg)
 {
-	struct dtx_ult_arg	  *ult_arg = arg;
-	struct dtx_leader_handle  *dlh = ult_arg->dlh;
-	ABT_future		  future = dlh->dlh_future;
-	uint32_t		  i;
-	int			  rc = 0;
+	struct dtx_ult_arg		*ult_arg = arg;
+	struct dtx_leader_handle	*dlh = ult_arg->dlh;
+	struct dtx_sub_status		*sub;
+	ABT_future			 future = dlh->dlh_future;
+	uint32_t			 i;
+	int				 rc = 0;
 
 	D_ASSERT(future != ABT_FUTURE_NULL);
 	for (i = 0; i < dlh->dlh_sub_cnt; i++) {
-		struct dtx_sub_status *sub = &dlh->dlh_subs[i];
-
+		sub = &dlh->dlh_subs[i];
 		sub->dss_result = 0;
+		sub->dss_comp = 0;
```

Comment:
Not sure I follow the patch here.

if the segfault happen here, it must mean there are still sub th(future) are not being set yet, then why does dtx_leader_end() can pass the wait and free the doh()?

or I miss sth?

Reply:
Not such case. The ABT future compartment is set by the RPC callback, that is asynchronously. So when the last compartment is set via ABT_future_set(), the IO forward ULT may be still accessing the DTX handle. At that time, the main IO handler ULT is waken up on another CPU core and freed the DTX handle. That will cause the IO forward ULT to segment fault.

Reply:
In fact, such issue exist there at the beginning when we introduce the IO forward ULT years ago. But at that time, there was no CPU yield for the IO forward ULT, so such race window is very short, that is why we did not hit the failure. But recently, we introduce CPU yield for IO forward ULT to avoid block others for too long time, that increased such race windows. Then we have more chance to hit such segment fault.

Reply:
yes, but if there are still ult_arg->func() is not executed yet, i.e. future is not set yet, then the dtx_leader_end() should wait()?

Reply:
ult_arg->func() is executed, but the forwarded ULT will still access DTX handle, such as for the for() boundary check.

Reply:
Hmm,  I think not all ult_arg->func are executed. 


Thread 1 (Thread 0x7f3395b91700 (LWP 157230)):
2#0  dtx_leader_exec_ops_ult (arg=0x7f2d687bfd30) at src/dtx/dtx_common.c:1710
3        sub = 0x100000828
4        ult_arg = 0x7f2d687bfd30
5        dlh = 0x7f33a7beb970
6        future = 0x7f2d6908ba00
7        i = 87
8        rc = 0
9        __func__ = "dtx_leader_exec_ops_ult"
10        __PRETTY_FUNCTION__ = "dtx_leader_exec_ops_ult"
11#1  0x00007f34ad648ece in ABTD_ythread_func_wrapper () from /lib64/libabt.so.1
12No symbol table info available.
13#2  0x00007f34ad649071 in make_fcontext () from /lib64/libabt.so.1
14No symbol table info available.
15#3  0x0000000000000000 in ?? ()
16No symbol table info available.
17No core file now.

here is the stack trace. I think there are still ult_arg->func are not executed yet?

Reply:
ah, there is yield inside this function. I see.

Reply:
>But at that time, there was no CPU yield for the IO forward ULT, so such race window is very short, that is why we did not hit the failure. 

adding yield here is a bit strange. probably move this yield to cpd forward?